sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -1247,33 +1315,107 @@ public void startBundle(StartBundleContext c) throws IOException {
+        KV<ByteString, Iterable<Mutation>> record, BoundedWindow window) {
+      return (MutateRowResponse result, Throwable exception) -> {
+        if (exception != null) {
+          if (exception instanceof NotFoundException
+              && !((NotFoundException) exception).isRetryable()) {
+            // This case, of being an NotFoundException and not retryable,
+            // indicates an issue with the data. However, we can't know if this record
+            // failed, or if it was another record in the batch. Thus, we retry each record
+            // individually.
+            retryIndividualRecord(record, window);
+          } else {
+            failures.add(new BigtableWriteException(record, exception));
+          }
+        }
+      };
+    }
+
+    public void retryIndividualRecord(
+        KV<ByteString, Iterable<Mutation>> record, BoundedWindow window) {
+      try {
+        unbatchedWrites.add(
+            bigtableWriter
+                .writeRecordWithoutBatching(record)
+                .whenComplete(
+                    (singleMutationResult, singleException) -> {
+                      if (singleException != null) {
+                        if (singleException instanceof NotFoundException
+                            && !((NotFoundException) singleException).isRetryable()) {
+                          // if we get another NotFoundException, we know this is the bad record.
+                          badRecords.add(
+                              KV.of(new BigtableWriteException(record, singleException), window));
+                        } else {
+                          failures.add(new BigtableWriteException(record, singleException));
+                        }
+                      }
+                    }));
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
     @FinishBundle
     public void finishBundle(FinishBundleContext c) throws Exception {
       try {
+
         if (bigtableWriter != null) {
-          bigtableWriter.close();
+          try {
+            bigtableWriter.close();
+          } catch (IOException e) {
+            // If the writer fails due to a batching exception, but no failures were detected
+            // it means that error handling was enabled, and that errors were detected and routed
+            // to the error queue. Bigtable will successfully write other failures in the batch,
+            // so this exception should be ignored
+            if (!(e.getCause() instanceof BatchingException)) {
+              throw e;
+            }
+          }
           bigtableWriter = null;
         }
 
+        for (CompletionStage<MutateRowResponse> unbatchedWrite : unbatchedWrites) {
+          try {
+            unbatchedWrite.toCompletableFuture().get();
+          } catch (Exception e) {
+            // ignore exceptions here, they are handled by the .whenComplete as part of the
```

Comment:
This catches all exceptions. Is it possible to reduce the scope to some known exceptions (like IOException-BatchingException above) or its not necessary?

Reply:
Its not necessary. This get exists just to force completion, and the exceptions are already handled.